pkg/storage/ingest/fetcher_test.go

```diff
@@ -1054,12 +1061,176 @@ func TestConcurrentFetchers(t *testing.T) {
+		for i := 0; i < totalProducedRecords; i++ {
+			produceRecord(ctx, t, client, topicName, partitionID, recordValue)
+		}
+
+		// Create fetchers with tracking of uncompressed bytes
+		fetchers := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, recordsPerFetch, maxInflightBytes)
+
+		waitForStableBufferedRecords := func() {
+			previousBufferedRecords := int64(0)
+			assert.Eventually(t, func() bool {
+				bufferedRecords := fetchers.BufferedRecords()
+				stabilized := bufferedRecords == previousBufferedRecords
+				previousBufferedRecords = bufferedRecords
+				return stabilized
+			}, 2*time.Second, 100*time.Millisecond)
+		}
+		// Wait for buffered records to stabilize, we expect that they stabilize because the limit is in effect.
+		waitForStableBufferedRecords()
+
+		// Assert that we don't buffer more than maxInflightBytes
+		assert.LessOrEqualf(t, fetchers.BufferedBytes(), int64(maxInflightBytes), "Should not buffer more than %d bytes of records", maxInflightBytes)
+
+		// Consume one batch of records
+		fetches, _ := fetchers.PollFetches(ctx)
+		totalConsumedRecords := fetches.NumRecords()
+		require.Greater(t, totalConsumedRecords, 0, "Should have received some records")
+
+		// Allow time for more fetches
+		waitForStableBufferedRecords()
+
+		// Assert again that buffered bytes remain under limit
+		assert.LessOrEqualf(t, fetchers.BufferedRecords(), int64(maxInflightBytes), "Should still not buffer more than %d bytes after consuming some records", maxInflightBytes)
+
+		// Consume all remaining records and verify total
+		for totalConsumedRecords < totalProducedRecords {
+			fetches, _ = fetchers.PollFetches(ctx)
+			totalConsumedRecords += fetches.NumRecords()
+		}
+
+		// Allow time for more fetches
+		waitForStableBufferedRecords()
+
+		pollFetchesAndAssertNoRecords(t, fetchers)
+		assert.Equal(t, totalProducedRecords, totalConsumedRecords, "Should have received all records eventually")
+	})
+
+	t.Run("respect maximum buffered bytes limit with varying record sizes", func(t *testing.T) {
+		// This test makes sure that the buffer doesn't become inefficient when the size estimations change (from large records we switch to small records).
+		t.Parallel()
+
+		const (
+			topicName        = "test-topic"
+			partitionID      = 1
+			concurrency      = 30
+			recordsPerFetch  = 100
+			maxInflightBytes = 5_000_000
+
+			largeRecordsCount = 100
+			largeRecordSize   = 100_000
+			smallRecordsCount = 10_000
+			smallRecordSize   = 1000
+		)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel()
+
+		_, clusterAddr := testkafka.CreateCluster(t, partitionID+1, topicName)
+		client := newKafkaProduceClient(t, clusterAddr)
+
+		// Create fetchers early to ensure we don't miss any records
+		fetchers := createConcurrentFetchers(ctx, t, client, topicName, partitionID, 0, concurrency, recordsPerFetch, maxInflightBytes)
+
+		waitForStableBufferedRecords := func() {
```

Comment:
[nit] Since this function is used in 2 test cases, you could move it at the beginning of `TestConcurrentFetchers`.